src/storage/src/types/sources.rs

```diff
@@ -1523,25 +1528,119 @@ impl Generator {
+
+    pub fn views(&self) -> Vec<(&str, Vec<(&str, &str)>)> {
         match self {
-            Self::Counter => Row::pack_slice(&[Datum::Int64(offset.offset as i64)]),
+            Generator::Auction => {
+                vec![
+                    (
+                        "auctions",
+                        vec![
+                            ("id", "INT8"),
+                            ("item", "STRING"),
+                            ("end_time", "TIMESTAMPTZ"),
+                        ],
+                    ),
+                    (
+                        "bids",
+                        vec![
+                            ("id", "INT8"),
+                            ("auction_id", "INT8"),
+                            ("amount", "INT4"),
+                            ("bid_time", "TIMESTAMPTZ"),
+                        ],
+                    ),
+                ]
+            }
+            Generator::Counter => Vec::new(),
+        }
+    }
+
+    pub fn by_seed(&self, now: NowFn, seed: Option<u64>) -> GeneratorIterator {
+        GeneratorIterator {
+            generator: self.clone(),
+            invocation: 0,
+            pending: VecDeque::new(),
+            rng: SmallRng::seed_from_u64(seed.unwrap_or_default()),
+            now,
         }
     }
 }
 
-impl FromStr for Generator {
-    type Err = String;
+pub struct GeneratorIterator {
+    generator: Generator,
+    // Number of times `pending` has been refilled.
+    invocation: i64,
+    pending: VecDeque<Row>,
+    rng: SmallRng,
+    now: NowFn,
+}
 
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        match s {
-            "counter" => Ok(Self::Counter),
-            _ => Err("unknown load generator".into()),
+impl Iterator for GeneratorIterator {
+    type Item = Row;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.pending.is_empty() {
+            self.invocation += 1;
+            let pending = match self.generator {
+                Generator::Counter => {
+                    VecDeque::from([Row::pack_slice(&[Datum::Int64(self.invocation)])])
+                }
+                Generator::Auction => {
+                    const AUCTIONS: &[&str] = &[
+                        "Signed Memorabilia",
+                        "City Bar Crawl",
+                        "Best Pizza in Town",
+                        "Gift Basket",
+                        "Custom Art",
+                    ];
+
+                    let now = to_datetime((self.now)());
+                    let mut auction = Row::with_capacity(2);
+                    let mut packer = auction.packer();
+                    packer.push(Datum::String("auctions"));
+                    packer.push_list(&[
+                        Datum::String(&self.invocation.to_string()), // auction id
+                        Datum::String(AUCTIONS.choose(&mut self.rng).unwrap()), // item
+                        Datum::String(&(now + chrono::Duration::seconds(10)).to_string()), // end time
```

Comment:
This just happens to produce a parseable representation, right? `strconv::format_timestamptz` seems like a safer choice.

Reply:
Bummer that you have to roundtrip through the string representation at all. Wish we had a proper union type.